tracing-appender/src/rolling.rs

```diff
@@ -327,17 +334,155 @@ impl Rotation {
-        match *self {
-            Rotation::MINUTELY => format!("{}.{}", filename, date.format("%F-%H-%M")),
-            Rotation::HOURLY => format!("{}.{}", filename, date.format("%F-%H")),
-            Rotation::DAILY => format!("{}.{}", filename, date.format("%F")),
-            Rotation::NEVER => filename.to_string(),
+/// A template that can be used to get the name of the next log file.
+pub trait FilenameTemplate: Send + Sync + 'static {
+    /// Get the filename that corresponds to this timestamp.
+    fn next_log_file(&mut self, date: &DateTime<Utc>, rotation: &Rotation) -> PathBuf;
+}
+
+impl<F> FilenameTemplate for F
+where
+    F: FnMut(&DateTime<Utc>, &Rotation) -> PathBuf + Send + Sync + 'static,
+{
+    fn next_log_file(&mut self, date: &DateTime<Utc>, rotation: &Rotation) -> PathBuf {
+        (self)(date, rotation)
+    }
+}
+
+impl FilenameTemplate for Box<dyn FilenameTemplate + Send + Sync + 'static> {
+    fn next_log_file(&mut self, date: &DateTime<Utc>, rotation: &Rotation) -> PathBuf {
+        (**self).next_log_file(date, rotation)
+    }
+}
+
+/// Get a [`FilenameTemplate`] with a timestamp appended to a common prefix.
+pub fn prefixed<D, P>(log_directory: D, prefix: P) -> Prefixed
+where
+    D: AsRef<Path>,
+    P: AsRef<OsStr>,
+{
+    Prefixed::new(log_directory, prefix)
+}
+
+/// A [`FilenameTemplate`] with a timestamp appended to a common prefix.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Prefixed {
+    log_directory: PathBuf,
+    prefix: OsString,
+}
+
+impl Prefixed {
+    /// Create a new [`Prefixed`] `FileTemplate`.
+    pub fn new<D, P>(log_directory: D, prefix: P) -> Self
+    where
+        D: AsRef<Path>,
+        P: AsRef<OsStr>,
+    {
+        Prefixed {
+            log_directory: log_directory.as_ref().to_owned(),
+            prefix: prefix.as_ref().to_owned(),
         }
     }
 }
 
+impl FilenameTemplate for Prefixed {
+    fn next_log_file(&mut self, date: &DateTime<Utc>, rotation: &Rotation) -> PathBuf {
+        let mut last_segment = self.prefix.clone();
+
+        if let Some(timestamp) = format_date_for_rotation(date, rotation) {
+            last_segment.push(".");
+            last_segment.push(&timestamp);
+        }
+
+        self.log_directory.join(last_segment)
+    }
+}
+
+/// Create a new [`FilenameTemplate`] which uses a common file name and
+/// extension.
+///
+/// # Examples
+///
+/// ```rust
+/// # use chrono::prelude::*;
+/// # use tracing_appender::rolling::{Rotation, FilenameTemplate};
+/// # use std::path::Path;
+/// let mut template = tracing_appender::rolling::with_name_and_extension(
+///     "/var/log",
+///     "MyApplication",
+///     "log",
+/// );
+///
+/// let date = Utc.ymd(2020, 2, 1).and_hms(12, 20, 15);
+/// let rotation = Rotation::DAILY;
+///
+/// let filename = template.next_log_file(&date, &rotation);
+/// assert_eq!(filename, Path::new("/var/log/MyApplication.2020-02-01.log"));
```

Comment:
nit: Prefer this to be in a unit test

Reply:
My thinking was this lets people see how the `FilenameTemplate` is created and used, and the output it generates. 

Although I guess I could just say that in the doc-comments, couldn't I?

Reply:
Yeah, as commented further down, we should still have a comment on how to use this. I just prefered to have the tests in the test section.